### PR TITLE
Add command for persistent job handler when msf restart

### PR DESCRIPTION
**Pull request by Green-m**
~## Working in process~

~Add the command `persist`~
According to https://github.com/rapid7/metasploit-framework/projects/6 to update the command `jobs` to support for persistent jobs when console restart.
```
Enable the ability to create a handler description that's persistent, so that when MSF restarts those handlers come alive again.

This is currently "emulated" via the CLI using -r blerp.rc.
```

I know there are many problems to be solved.
I'm glad to receive some suggestions.

## Verification(updated)

```
msf5 exploit(multi/handler) > jobs 

Jobs
====

  Id  Name                    Payload                          Payload opts
  --  ----                    -------                          ------------
  1   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.136:4444

msf5 exploit(multi/handler) > jobs -P
Making all jobs persistent ...
Added job 1 as a persistent_job.
msf5 exploit(multi/handler) > jobs -l

Jobs
====

  Id  Name                    Payload                          Payload opts
  --  ----                    -------                          ------------
  1   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.136:4444

msf5 exploit(multi/handler) > jobs -l -v

Jobs
====

  Id  Name                    Payload                          Payload opts               URIPATH  Start Time                 Handler opts  Persist
  --  ----                    -------                          ------------               -------  ----------                 ------------  -------
  1   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.136:4444           2018-08-07 05:10:41 -0400                true

msf5 exploit(multi/handler) > kill 1 
[*] Stopping the following job(s) and remove their persistence: 1
[*] Stopping job 1
msf5 exploit(multi/handler) > jobs -l -v

Jobs
====

No active jobs.

msf5 exploit(multi/handler) > exploit -j
[*] Exploit running as background job 2.

[*] Started reverse TCP handler on 192.168.77.136:4444 
msf5 exploit(multi/handler) > jobs -l -v

Jobs
====

  Id  Name                    Payload                          Payload opts               URIPATH  Start Time                 Handler opts  Persist
  --  ----                    -------                          ------------               -------  ----------                 ------------  -------
  2   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.136:4444           2018-08-07 05:35:59 -0400                false

```

## Verification(obsolete)

List the steps needed to make sure this thing works

```
[*] Processing ../multi_handler.txt for ERB directives.
resource (../multi_handler.txt)> use exploit/multi/handler
resource (../multi_handler.txt)> set lhost 192.168.77.141
lhost => 192.168.77.141
resource (../multi_handler.txt)> set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
resource (../multi_handler.txt)> set exitonsession false
exitonsession => false
msf5 exploit(multi/handler) > exploit -j
[*] Exploit running as background job 0.

[*] Started reverse TCP handler on 192.168.77.141:4444 
msf5 exploit(multi/handler) > jobs 

Jobs
====

  Id  Name                    Payload                          Payload opts
  --  ----                    -------                          ------------
  0   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.141:4444

msf5 exploit(multi/handler) > persist 

Persistent Jobs
===============

No persistent jobs.

msf5 exploit(multi/handler) > persist -a 0 
Added job 0 as a persistent_job.
msf5 exploit(multi/handler) > persist 

Persistent Jobs
===============

exploit/multi/handler - windows/meterpreter/reverse_tcp - 192.168.77.141:4444

msf5 exploit(multi/handler) > exit

$ ./msfconsole                        
[-] ***rting the Metasploit Framework console...|
[-] * WARNING: No database support: No database YAML file
[-] ***
                                                 

       =[ metasploit v5.0.0-dev-0145744255                ]
+ -- --=[ 1778 exploits - 1011 auxiliary - 308 post       ]
+ -- --=[ 538 payloads - 41 encoders - 10 nops            ]
+ -- --=[ ** This is Metasploit 5 development branch **   ]

[*] Starting persistent handler...
msf5 > jobs 

Jobs
====

  Id  Name                    Payload                          Payload opts
  --  ----                    -------                          ------------
  0   Exploit: multi/handler  windows/meterpreter/reverse_tcp  tcp://192.168.77.141:4444

msf5 > 

```

### CODE REVIEW

**Comment by jmartin-r7** [comment]
This is a nice first pass!  Thanks for getting the ball rolling.

An idea to adjusting the interaction, consider making this an option of the `jobs` command vs a separate thing to manage.  For me the functionality applies specifically to things that are already contained in the `jobs` list, so I am thinking something like:
```
msf > jobs -h
Usage: jobs [options]

Active job manipulation and interaction.

OPTIONS:

    -K        Terminate all running jobs.
    -S <opt>  Row search filter.
    -P        Persist all running jobs on restart.
    -h        Help banner.
    -i <opt>  Lists detailed information about a running job.
    -k <opt>  Terminate jobs by job ID and/or range.
    -l        List all running jobs.
    -p        Persist job ID on restart.
    -v        Print more detailed info.  Use with -i and -l

msf exploit(multi/handler) > jobs

Jobs
====

  Id  Name                    Payload                               Payload opts
  --  ----                    -------                               ------------
  0   Exploit: multi/handler  linux/mipsbe/meterpreter_reverse_tcp  tcp://127.0.0.1:4444

msf exploit(multi/handler) > jobs -p 0
[*] Job 0 restored on console restarted until specifically terminated.
```
This may make the code expand a little since it re-uses the `-k` and `-K` options for removal of persisted jobs, on the other hand it may also you to encapsulate the code more as well.

Would love to see other's input on this as well, either approach has merits.

**Comment by Green-m** [comment]
@jmartin-r7 
Thanks for your advice, it seems reasonable, I will update it soon.

**Comment by Green-m** [comment]
@jmartin-r7 
Some questions about to merge command `persist` to `job`: 

1. To list all persistent jobs rather than running jobs, it's better to expand command `job -l -v` or add a new option like `job --persist-list` ?

2. When we kill a running job with `job -k x`, should persistent job be killed too? Is there a situation that we want it to be seperate?

**Comment by Green-m** [comment]
@jmartin-r7 
Sorry for the late. Now I have updated it, please check it out.

**Review by jmartin-tech — COMMENTED** [review]
Thanks for getting this refactored,  I like this user experience.  

I would like to see a more standard and readable format to the persistence file as noted in my comments.

**Comment by Green-m** [comment]
@jmartin-r7 
I really thank you for your helpful suggestion. I am not good at code format and ruby style, so I know that code in this PR is not readable and lucid. 
I will update it next weekday : )

**Comment by Green-m** [comment]
@jmartin-r7 
I have updated it. Please check it again.
And I just changed some global `rescue` code, not for all, maybe cleaner.

**Review by jmartin-tech — COMMENTED** [review]
Style change requests.

File: lib/msf/base/serializer/readable_text.rb [hunk 1]
```diff
@@ -766,6 +766,11 @@ def self.dump_jobs(framework, verbose = false, indent = DefaultIndent, col = Def
       'Columns' => columns
       )
 
+    # Get the persistent job info.
+    if verbose
+      persist_list = JSON.parse(File.read(Msf::Config.persist_file)) rescue []
```
[the remark this answers is not in the document]
Comment (Green-m) [reply lib/msf/base/serializer/readable_text.rb]: Updated, thanks.

**Comment by timwr** [comment]
Excellent work everyone :)

File: lib/msf/ui/console/driver.rb [hunk 2]
```diff
@@ -178,6 +178,22 @@ def initialize(prompt = DefaultPrompt, prompt_char = DefaultPromptChar, opts = {
...
     end
 
+    # Process persistent job handler
+    begin
+      restore_handlers = JSON.parse(File.read(Msf::Config.persist_file))
+    rescue Errno::ENOENT, JSON::ParserError
+      restore_handlers = nil
+    end
+
+    unless restore_handlers.nil?
```
Comment (wvu-r7) [inline lib/msf/ui/console/driver.rb]: For next time, double negatives can be confusing. I think `if restore_handlers` would suffice.

**Comment by tdoan-r7** [comment]
# Release Notes
Support has been added to the command jobs for persistent jobs.

**Pull request merged**